File: server_flask/app/utils/slugify.py
import re
import unicodedata
import uuid
from app.models import Post


def generate_unique_slug(title, db_session):
    # Normalize and remove unwanted characters
    base_slug = unicodedata.normalize("NFKD", title)
    base_slug = re.sub(r"[^\w\s-]", "", base_slug.lower())
    base_slug = re.sub(r"[\s_-]+", "-", base_slug).strip("-")

    # Fallback in case base_slug is empty (e.g., title is just "?" or emoji)
    if not base_slug:
        base_slug = f"post-{uuid.uuid4().hex[:8]}"

    slug = base_slug
    counter = 1

    while db_session.query(Post).filter_by(slug=slug).first() is not None:
        slug = f"{base_slug}-{counter}"
        counter += 1

    return slug

File: server_node/utils/slugify.js
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();

const slugify = (title) => {
  // Sanitize and remove unwanted characters from post title to create a URL friendly slug
  let baseSlug = title
    .normalize("NFKD")
    .replace(/[^\w\s-]/g, "")
    .toLowerCase()
    .trim()
    .replace(/[\s_-]+/g, "-");

  // If baseSlug is empty, generate a random slug
  if (!baseSlug) {
    baseSlug = `post-${Math.random().toString(36).substring(2, 10)}`;
  }

  return baseSlug;
};

const generateUniqueSlug = async (title) => {
  const baseSlug = slugify(title);
  let slug = baseSlug;
  let counter = 1;

  while (await prisma.post.findUnique({ where: { slug } })) {
    slug = `${baseSlug}-${counter}`;
    counter++;
  }

  return slug;
};

module.exports = { generateUniqueSlug };
